Don't use private traits in public APIs

For example, hyper master currently has this:

```rust
impl<I, B, S, E> Future for Connection<I, S, E>
where
    S: HttpService<Body, ResBody=B>,
    S::Error: Into<Box<dyn StdError + Send + Sync>>,
    I: AsyncRead + AsyncWrite + Unpin + 'static,
    B: Payload + 'static,
    B::Data: Unpin,
    E: H2Exec<S::Future, B>,
```

`HttpService` and `H2Exec` are both private traits, though. This makes it a huge pain for code working generically with connections to express the proper bounds in their code. You basically have to try to reverse engineer the impl bounds by bouncing off the compiler :(

The H2Exec bound in particular is a bit of a nightmare:

```
error[E0599]: no method named `graceful_shutdown` found for type `std::pin::Pin<&mut hyper::server::conn::Connection<I, server::idle::TrackedService<S>>>` in the current scope
   --> sorcery/src/server/idle.rs:163:19
    |
163 |         this.conn.graceful_shutdown();
    |                   ^^^^^^^^^^^^^^^^^ method not found in `std::pin::Pin<&mut hyper::server::conn::Connection<I, server::idle::TrackedService<S>>>`
    |
    = note: the method `graceful_shutdown` exists but the following trait bounds were not satisfied:
            `hyper::common::exec::Exec : hyper::common::exec::H2Exec<server::idle::TrackedFuture<<S as tower_service::Service<http::request::Request<hyper::body::body::Body>>>::Future>, _>`
```

when using `impl Service` for a client param, I get:
```
error[E0599]: no method named `request` found for type `impl hyper::service::Service` in the current scope
  --> src/reddit.rs:39:22
   |
39 |     let res = client.request(req).await?;
   |                      ^^^^^^^ method not found in `impl hyper::service::Service`
```
Although, I haven't finished going through all the compiler errors yet.

so, I'm pretty stuck at this point.

If I use `impl Service` as a type in the fn parameters, I get an error like:
```
error[E0277]: the trait bound `&hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>: tower_service::Service<http::request::Request<hyper::body::body::Body>>` is not satisfied
  --> src/main.rs:31:21
   |
31 |     let new_posts = fetch_reddit_new(&client, subreddit, n).await?;
   |                     ^^^^^^^^^^^^^^^^ the trait `tower_service::Service<http::request::Request<hyper::body::body::Body>>` is not implemented for `&hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>`
   | 
  ::: src/reddit.rs:15:14
   |
15 | pub async fn fetch_reddit_new(
   |              ----------------
16 |     client: impl Service<Request<hyper::Body>>,
   |                  ----------------------------- required by this bound in `reddit::fetch_reddit_new`
   |
   = help: the following implementations were found:
             <hyper::client::Client<C, B> as tower_service::Service<http::request::Request<B>>>
```
If I use `Client` as a parameter's type, I have a bunch of bounds I need to add, but get stuck on needing `Connect`, which is sealed.

Since it was pretty straightforward to have a `Client` generic over `Connect` in the past, it's a little surprising that it's not possible now (or requires some indirection I don't know yet).

Really looking forward to upgrading to hyper 0.13, so help figuring this out is definitely appreciated.

--- a/src/common/exec.rs
+++ b/src/common/exec.rs
@@ -3,16 +3,7 @@ use std::future::Future;
 use std::pin::Pin;
 use std::sync::Arc;
 
-#[cfg(feature = "server")]
-use crate::body::Body;
-#[cfg(all(feature = "http2", feature = "server"))]
-use crate::proto::h2::server::H2Stream;
 use crate::rt::Executor;
-
-#[cfg(feature = "server")]
-pub trait ConnStreamExec<F, B: Body>: Clone {
-    fn execute_h2stream(&mut self, fut: H2Stream<F, B>);
-}
 
 pub(crate) type BoxSendFuture = Pin<Box<dyn Future<Output = ()> + Send>>;
 
@@ -46,31 +37,6 @@ impl Exec {
 impl fmt::Debug for Exec {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("Exec").finish()
-    }
-}
-
-#[cfg(feature = "server")]
-impl<F, B> ConnStreamExec<F, B> for Exec
-where
-    H2Stream<F, B>: Future<Output = ()> + Send + 'static,
-    B: Body,
-{
-    fn execute_h2stream(&mut self, fut: H2Stream<F, B>) {
-        self.execute(fut)
-    }
-}
-
-// ==== impl Executor =====
-
-#[cfg(feature = "server")]
-impl<E, F, B> ConnStreamExec<F, B> for E
-where
-    E: Executor<H2Stream<F, B>> + Clone,
-    H2Stream<F, B>: Future<Output = ()>,
-    B: Body,
-{
-    fn execute_h2stream(&mut self, fut: H2Stream<F, B>) {
-        self.execute(fut)
     }
 }
 

--- a/src/proto/h2/server.rs
+++ b/src/proto/h2/server.rs
@@ -13,7 +13,7 @@ use tracing::{debug, trace, warn};
 
 use super::{ping, PipeToSendStream, SendBuf};
 use crate::body::{Body, Incoming as IncomingBody};
-use crate::common::exec::ConnStreamExec;
+use crate::rt::bounds::Http2ConnExec;
 use crate::common::time::Time;
 use crate::common::{date, task, Future, Pin, Poll};
 use crate::ext::Protocol;
@@ -110,7 +110,7 @@ where
     S: HttpService<IncomingBody, ResBody = B>,
     S::Error: Into<Box<dyn StdError + Send + Sync>>,
     B: Body + 'static,
-    E: ConnStreamExec<S::Future, B>,
+    E: Http2ConnExec<S::Future, B>,
 {
     pub(crate) fn new(
         io: T,
@@ -186,7 +186,7 @@ where
     S: HttpService<IncomingBody, ResBody = B>,
     S::Error: Into<Box<dyn StdError + Send + Sync>>,
     B: Body + 'static,
-    E: ConnStreamExec<S::Future, B>,
+    E: Http2ConnExec<S::Future, B>,
 {
     type Output = crate::Result<Dispatched>;
 
@@ -240,7 +240,7 @@ where
     where
         S: HttpService<IncomingBody, ResBody = B>,
         S::Error: Into<Box<dyn StdError + Send + Sync>>,
-        E: ConnStreamExec<S::Future, B>,
+        E: Http2ConnExec<S::Future, B>,
     {
         if self.closing.is_none() {
             loop {

--- a/src/rt/bounds.rs
+++ b/src/rt/bounds.rs
@@ -1,4 +1,7 @@
-//! Sealed traits
+//! Trait aliases
+//!
+//! Traits in this module ease setting bounds and usually automatically
+//! implemented by implementing another trait.
 
 #[cfg(all(feature = "server", feature = "http2"))]
 pub use self::h2::Http2ConnExec;
@@ -12,7 +15,12 @@ mod h2 {
 
     /// An executor to spawn http2 connections.
     ///
+    /// This trait is implemented for any type that implements [`Executor`]
+    /// trait for any future.
+    ///
     /// This trait is sealed and cannot be implemented for types outside this crate.
+    ///
+    /// [`Executor`]: crate::rt::Executor
     pub trait Http2ConnExec<F, B: Body>: sealed::Sealed<(F, B)> + Clone {
         #[doc(hidden)]
         fn execute_h2stream(&mut self, fut: H2Stream<F, B>);

--- a/src/rt/bounds.rs
+++ b/src/rt/bounds.rs
@@ -1,0 +1,61 @@
+//! Sealed traits
+
+#[cfg(all(feature = "server", feature = "http2"))]
+pub use self::h2::Http2ConnExec;
+
+#[cfg(all(feature = "server", feature = "http2"))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "server", feature = "http2"))))]
+mod h2 {
+    use crate::{common::exec::Exec, proto::h2::server::H2Stream, rt::Executor};
+    use http_body::Body;
+    use std::future::Future;
+
+    /// An executor to spawn http2 connections.
+    ///
+    /// This trait is sealed and cannot be implemented for types outside this crate.
+    pub trait Http2ConnExec<F, B: Body>: sealed::Sealed<(F, B)> + Clone {
+        #[doc(hidden)]
+        fn execute_h2stream(&mut self, fut: H2Stream<F, B>);
+    }
+
+    impl<F, B> Http2ConnExec<F, B> for Exec
+    where
+        H2Stream<F, B>: Future<Output = ()> + Send + 'static,
+        B: Body,
+    {
+        fn execute_h2stream(&mut self, fut: H2Stream<F, B>) {
+            self.execute(fut)
+        }
+    }
+
+    impl<F, B> sealed::Sealed<(F, B)> for Exec
+    where
+        H2Stream<F, B>: Future<Output = ()> + Send + 'static,
+        B: Body,
+    {
+    }
+
+    #[doc(hidden)]
+    impl<E, F, B> Http2ConnExec<F, B> for E
+    where
+        E: Executor<H2Stream<F, B>> + Clone,
+        H2Stream<F, B>: Future<Output = ()>,
+        B: Body,
+    {
+        fn execute_h2stream(&mut self, fut: H2Stream<F, B>) {
+            self.execute(fut)
+        }
+    }
+
+    impl<E, F, B> sealed::Sealed<(F, B)> for E
+    where
+        E: Executor<H2Stream<F, B>> + Clone,
+        H2Stream<F, B>: Future<Output = ()>,
+        B: Body,
+    {
+    }
+
+    mod sealed {
+        pub trait Sealed<T> {}
+    }
+}

--- a/src/rt/mod.rs
+++ b/src/rt/mod.rs
@@ -14,6 +14,26 @@ use std::{
 };
 
 /// An executor of futures.
+///
+/// This trait should be implemented for any future.
+///
+/// # Example
+///
+/// ```
+/// # use hyper::rt::Executor;
+/// # use std::future::Future;
+/// struct TokioExecutor;
+///
+/// impl<F> Executor<F> for TokioExecutor
+/// where
+///     F: Future + Send + 'static,
+///     F::Output: Send + 'static,
+/// {
+///     fn execute(&self, future: F) {
+///         tokio::spawn(future);
+///     }
+/// }
+/// ```
 pub trait Executor<Fut> {
     /// Place the future into the executor to be run.
     fn execute(&self, fut: Fut);

--- a/src/rt/mod.rs
+++ b/src/rt/mod.rs
@@ -5,6 +5,8 @@
 //! If the `runtime` feature is disabled, the types in this module can be used
 //! to plug in other runtimes.
 
+pub mod bounds;
+
 use std::{
     future::Future,
     pin::Pin,

--- a/src/server/conn/http2.rs
+++ b/src/server/conn/http2.rs
@@ -9,9 +9,9 @@ use pin_project_lite::pin_project;
 use tokio::io::{AsyncRead, AsyncWrite};
 
 use crate::body::{Body, Incoming as IncomingBody};
-use crate::common::exec::ConnStreamExec;
 use crate::common::{task, Future, Pin, Poll, Unpin};
 use crate::proto;
+use crate::rt::bounds::Http2ConnExec;
 use crate::service::HttpService;
 use crate::{common::time::Time, rt::Timer};
 
@@ -54,7 +54,7 @@ where
     I: AsyncRead + AsyncWrite + Unpin,
     B: Body + 'static,
     B::Error: Into<Box<dyn StdError + Send + Sync>>,
-    E: ConnStreamExec<S::Future, B>,
+    E: Http2ConnExec<S::Future, B>,
 {
     /// Start a graceful shutdown process for this connection.
     ///
@@ -78,7 +78,7 @@ where
     I: AsyncRead + AsyncWrite + Unpin + 'static,
     B: Body + 'static,
     B::Error: Into<Box<dyn StdError + Send + Sync>>,
-    E: ConnStreamExec<S::Future, B>,
+    E: Http2ConnExec<S::Future, B>,
 {
     type Output = crate::Result<()>;
 
@@ -259,7 +259,7 @@ impl<E> Builder<E> {
         Bd: Body + 'static,
         Bd::Error: Into<Box<dyn StdError + Send + Sync>>,
         I: AsyncRead + AsyncWrite + Unpin,
-        E: ConnStreamExec<S::Future, Bd>,
+        E: Http2ConnExec<S::Future, Bd>,
     {
         let proto = proto::h2::Server::new(
             io,

--- a/src/server/conn/http2.rs
+++ b/src/server/conn/http2.rs
@@ -99,7 +99,10 @@ where
 impl<E> Builder<E> {
     /// Create a new connection builder.
     ///
-    /// This starts with the default options, and an executor.
+    /// This starts with the default options, and an executor which is a type
+    /// that implements [`Http2ConnExec`] trait.
+    ///
+    /// [`Http2ConnExec`]: crate::rt::bounds::Http2ConnExec
     pub fn new(exec: E) -> Self {
         Self {
             exec: exec,

--- a/src/service/mod.rs
+++ b/src/service/mod.rs
@@ -26,7 +26,7 @@ mod service;
 mod util;
 
 #[cfg(all(any(feature = "http1", feature = "http2"), feature = "server"))]
-pub(super) use self::http::HttpService;
+pub use self::http::HttpService;
 #[cfg(all(
     any(feature = "http1", feature = "http2"),
     any(feature = "server", feature = "client")

